### ISSUE TITLE
How to auto-login on WooCommerce after logging in to back office

### ISSUE THREAD

**Issue by ganthen** [daggerhart/openid-connect-generic#411]
Hello,

Our custom application uses IdentityServer for all security and logon scenario's for our custom app. However, the web site also relies on the same IdentityServer for its logon to the WooCommerce pages, hence the openid-connect client plugin. I am looking for assistance, lines of code, pointers to the proper webhook(s) or so to automatically log on my user on the web shop (both existing users as well as new registered user on our custom app) after having logged on to (the identityServer of) our app. I can rely on the presence of  a cookie. How can I add some lines of code to log in automatically on my WordPress web shop ? Anybody can help me out here? Do you have a code snippet to add, and where should it hook up ? 

For now, my users will first log in to the app, then navigate to the web shop and see the button "Login", having to click it manually first, and being confusing for the end user. He just came to fill in his login information, and now has to fill in the login fields all over again after clicking the login button.

Some help would be appreciated :-)

**Comment by charleslcso** [comment]
yes, we just hit the same issue.... can't figure out how to get WooCommerce to login too after logging into server. Seems no help here, as the plugin is designed solely for WP?

**Comment by timnolte** [comment]
So to be clear, this isn't really a WooCommerce issue. WooCommerce is just a plugin for WordPress and uses the WordPress login process for customer logins as well. Additionally, the fact that a user is having to authenticate with the IDP twice tells me that your IDP login setup isn't quite right. I use this plugin on a bunch of client WordPress sites where like 20 WordPress sites are all configured to use the same IDP. If I tell the IDP to keep me logged in then if I go to any of the other WordPress sites I don't have to login with the IDP using my username & password again.

**Comment by charleslcso** [comment]
Okok, so we are not setting it up properly. Can you share your experience on what we should look at first?

**Comment by timnolte** [comment]
@charleslcso so first of all what IDP are you using?

**Comment by charleslcso** [comment]
@timnolte we are using [Keycloak](https://www.keycloak.org)

**Comment by timnolte** [comment]
@charleslcso perfect I have a test Keycloak instance setup that I can use to double check the setup. Have you taken a look at our Wiki here that has some Keycloak setup instructions?

**Comment by charleslcso** [comment]
Your setup is/should help us tremendously. Looking forward to your reply!

**Comment by timnolte** [comment]
> May I ask... if I log out of WP/WC, would it sign out of server too?

@charleslcso so if your plugin setup is complete with a logout endpoint then yes, when logging out of WordPress/WooCommerce you should get logged out of your IDP(Keycloak) as well.

**Comment by timnolte** [comment]
@charleslcso FYI, I have a WooCommerce test site, since I also maintain a WooCommerce plugin/add-on, so I'm going to setup the OIDC plugin along with a client setup using our test Keycloak instance and do some testing.

**Comment by charleslcso** [comment]
> @charleslcso FYI, I have a WooCommerce test site, since I also maintain a WooCommerce plugin/add-on, so I'm going to setup the OIDC plugin along with a client setup using our test Keycloak instance and do some testing.

Great!

Re log in, it is unintuitive to add 1 extra trip to /wp-login before going back to Checkout page manually. Otherwise, a direct login/register (simply want to log in and not going to the Checkout path) will also head to Checkout! UX wise it is not what users expect.

I believe the cause is WP session not instantiated after login/register and going directly back to Checkout page. If OIDC plugin can't handle this, I believe Woo Checkout need to modified to check whether user is logged in before displaying the Checkout page content.

Re log out, we have an end point. After thinking deeper on how SSO works on log out, we opt to log out of each application individually.

**Comment by timnolte** [comment]
@charleslcso hmm, so I think the problem seems that you need to leverage the OIDC plugin's use of redirecting the user back to the source of the login flow by making sure that your WooCommerce templates are leveraging the shortcodes to produce a login button or URL(for a link) generated by our plugin. This will ensure that when users login at checkout that they are sent back to the checkout, however at the same time your WordPress admins & store managers get redirected back to the WordPress Dashboard if they are also using the Keycloak IDP for logins. Are you properly changing the WooCommerce templates or just using the default links for signing it that take users to the WordPress login screen?

**Comment by timnolte** [comment]
@charleslcso this is the documentation on the available shortcodes: https://github.com/oidc-wp/openid-connect-generic/wiki/Shortcodes

**Comment by charleslcso** [comment]
> @charleslcso hmm, so I think the problem seems that you need to leverage the OIDC plugin's use of redirecting the user back to the source of the login flow by making sure that your WooCommerce templates are leveraging the shortcodes to produce a login button or URL(for a link) generated by our plugin. This will ensure that when users login at checkout that they are sent back to the checkout, however at the same time your WordPress admins & store managers get redirected back to the WordPress Dashboard if they are also using the Keycloak IDP for logins. Are you properly changing the WooCommerce templates or just using the default links for signing it that take users to the WordPress login screen?

We added OIDC plugin at a very late stage. At the beginning we pick a theme that claim to be WC compatible. We haven't changed the theme code at all. 

We did use the plugin's referrer URL for going back to the referring page. However, at the stock setup the page will go into redirect loop, and hence it breaks everything. The workaround we use is to patch the "Proceed to Checkout" button to point directly to our OAuth server with a redirect_uri value.

May be we should go back to the redirect loop and figure out why it loops instead.

**Comment by timnolte** [comment]
Hmm, interesting. I'll work on getting my setup going as a demo to see if I can reproduce any of this. Thanks for the details.

**Comment by charleslcso** [comment]
> Hmm, interesting. I'll work on getting my setup going as a demo to see if I can reproduce any of this. Thanks for the details.

Just in case, the theme we use is GeneratePress Version: 3.1.3

**Comment by timnolte** [comment]
@charleslcso sorry for the delays. So I got my test WooCommerce site setup and configured it to be connected to my test Keycloak server. I also installed the GeneratePress theme. The next thing I did was created a child theme of GeneratePress so that I could create the required custom WooCommerce account login form template. This [custom WooCommerce template](https://woocommerce.com/document/template-structure/#how-to-edit-files) would be `<child-theme>/woocommerce/myaccount/form-login.php`. You can see the output [here](https://test.wclocalpickup.com/my-account/). The login form section of the template looks like this:
```php
<form class="woocommerce-form woocommerce-form-login login" action="#">

	<?php do_action( 'woocommerce_login_form_start' ); ?>

	<?php do_action( 'woocommerce_login_form' ); ?>

	<p class="form-row">
              <button type="button" class="woocommerce-button button woocommerce-form-login__submit" onclick="location.href='<?php echo do_shortcode( '[openid_connect_generic_auth_url]' ); ?>';"><?php esc_html_e( 'Log in', 'woocommerce' ); ?></button>
	</p>

	<?php do_action( 'woocommerce_login_form_end' ); ?>

</form>
```
I used our `[openid_connect_generic_auth_url]` shortcode in order to try and retain the general login button styling. We also have a button shortcode however I'm noticing now that is has some hardcoded styles intended for the WordPress login page that I think we need to do some work to allow some different types of buttons that allow better theme/site integration. The URL shortcode does also provide this flexibility.

So with this setup when I click on the login button I'm taken to the IDP, after I login there I am redirected back to my WooCommerce My Account page being logged in. However, to be clear I don't believe this actually speaks to the OP's post.
